### PR TITLE
Sk fix large projects

### PR DESCRIPTION
**Pull request by ShaneKilkelly**

### CODE REVIEW

File: app/coffee/Features/Cooldown/CooldownManager.coffee [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+
+module.exports = CooldownManager =
+
+	_buildKey: (projectId) ->
+		"Cooldown:{#{projectId}}"
+
+	putProjectOnCooldown: (projectId, callback=(err)->) ->
+		logger.log {projectId},
+			"[Cooldown] putting project on cooldown for #{COOLDOWN_IN_SECONDS} seconds"
```
Comment (henryoswald) [inline app/coffee/Features/Cooldown/CooldownManager.coffee]: this log line is on 2 lines

File: app/coffee/Features/Cooldown/CooldownMiddlewear.coffee [hunk 2]
```diff
@@ -0,0 +1,15 @@
+CooldownManager = require('./CooldownManager')
+logger = require('logger-sharelatex')
+
+
+module.exports = CooldownMiddlewear =
+
+	freezeProject: (req, res, next) ->
+			projectId = req.params.Project_id
```
Comment (henryoswald) [inline app/coffee/Features/Cooldown/CooldownMiddlewear.coffee]: worth null checking this Project_id

File: app/coffee/infrastructure/Mongoose.coffee [hunk 3]
```diff
@@ -2,6 +2,8 @@ mongoose = require('mongoose')
 Settings = require 'settings-sharelatex'
 logger = require('logger-sharelatex')
 
+mongoose.Promise = global.Promise
```
Comment (henryoswald) [inline app/coffee/infrastructure/Mongoose.coffee]: what is this Promise for?

File: package.json [hunk 4]
```diff
@@ -39,7 +39,7 @@
     "mimelib": "0.2.14",
     "mocha": "1.17.1",
     "mongojs": "2.4.0",
-    "mongoose": "4.1.0",
+    "mongoose": "4.9.8",
```
Comment (henryoswald) [inline package.json]: Is it worth making the cool down and mongoose upgrade separate deploys so if this PR fixes the issue we know if it was mongoose or the cool down code?

We will need to give the site a really good test with the new mongoose.

File: app/coffee/Features/Cooldown/CooldownManager.coffee [hunk 5]
```diff
@@ -0,0 +1,24 @@
+RedisWrapper = require('../../infrastructure/RedisWrapper')
+rclient = RedisWrapper.client('cooldown')
```
Comment (henryoswald) [inline app/coffee/Features/Cooldown/CooldownManager.coffee]: I've lost touch with how we do this, are we setting a different connection group for each feature?

Comment (jpallen) [inline app/coffee/Features/Cooldown/CooldownManager.coffee]: Yes, this is what I just merged. This bit is 👍

File: app/coffee/infrastructure/Mongoose.coffee (hunk 3 above)
Comment (ShaneKilkelly) [inline app/coffee/infrastructure/Mongoose.coffee]: The new version of mongoose issues a deprecation warning, about it's native Promises being deprecated and needing to set a promise implementation explicitly (http://mongoosejs.com/docs/promises.html)

File: package.json (hunk 4 above)
Comment (ShaneKilkelly) [inline package.json]: Good idea, will make a separate PR to upgrade mongoose

File: app/coffee/Features/Cooldown/CooldownManager.coffee (hunk 5 above)
Comment (ShaneKilkelly) [inline app/coffee/Features/Cooldown/CooldownManager.coffee]: Yup, different features can go to different connections

File: app/coffee/infrastructure/Mongoose.coffee (hunk 3 above)
Comment (henryoswald) [inline app/coffee/infrastructure/Mongoose.coffee]: right, is this why we need to change all the require("mongoose") paths?

Comment (ShaneKilkelly) [inline app/coffee/infrastructure/Mongoose.coffee]: Yeah, so that they get this promise implementation across the board.

Looking at it again, it makes more sense to do these two things separately as different PRs

File: app/coffee/Features/Cooldown/CooldownMiddlewear.coffee (hunk 2 above)
Comment (ShaneKilkelly) [inline app/coffee/Features/Cooldown/CooldownMiddlewear.coffee]: done
